supportedBackends/spark/src/main/scala/cromwell/backend/impl/spark/SparkJobExecutionActor.scala

```diff
@@ -0,0 +1,207 @@
+           case Finished => processSuccess(0)
+           case Failed(error: Throwable) =>  FailedNonRetryableResponse(jobDescriptor.key, error, None)
+        } recover {
+         case error: Throwable => FailedNonRetryableResponse(jobDescriptor.key, error, None)
+       }
+      case false => Future.successful(processSuccess(0))
+    }
+  }
+
+  private def processSuccess(rc: Int) = {
+    evaluateOutputs(callEngineFunction, outputMapper(jobPaths)) match {
+      case Success(outputs) => SucceededResponse(jobDescriptor.key, Some(rc), outputs)
+      case Failure(e) =>
+        val message = Option(e.getMessage) map {
+          ": " + _
+        } getOrElse ""
+        FailedNonRetryableResponse(jobDescriptor.key, new Throwable("Failed post processing of outputs" + message, e), Option(rc))
+    }
+  }
+
+  /**
+    * Abort a running job.
+    */
+  override def abort(): Unit = Future.failed(new UnsupportedOperationException("SparkBackend currently doesn't support aborting jobs."))
+
+  /**
+    *
+    */
+  private def createExecutionFolderAndScript(): Unit = {
+    try {
+      log.debug("{} Creating execution folder: {}", tag, executionDir)
+      executionDir.toString.toFile.createIfNotExists(true)
+
+      log.debug("{} Resolving job command", tag)
+      val command= localizeInputs(jobPaths.callRoot, docker = false, fileSystems, jobDescriptor.inputs) flatMap {
+        localizedInputs => call.task.instantiateCommand(localizedInputs, callEngineFunction, identity)
+      }
+
+      log.debug("{} Creating bash script for executing command: {}", tag, command)
+      //TODO: we should use shapeless Heterogeneous list here not good to have generic map
+      val attributes: Map[String, Any] = Map(
+        SparkCommands.APP_MAIN_CLASS -> runtimeAttributes.appMainClass,
+        SparkCommands.MASTER -> sparkMaster,
+        SparkCommands.EXECUTOR_CORES -> runtimeAttributes.executorCores,
+        SparkCommands.EXECUTOR_MEMORY -> runtimeAttributes.executorMemory.to(MemoryUnit.GB).amount.toLong,
+        SparkCommands.SPARK_APP_WITH_ARGS -> command.get,
+        SparkCommands.DEPLOY_MODE -> sparkDeployMode
+      )
+
+      val sparkSubmitCmd = cmds.sparkSubmitCommand(attributes)
+      val isSparkClusterMode = checkForSparkClusterMode(sparkDeployMode, sparkMaster)
+      val sparkCommand = if(isSparkClusterMode) {
+        sparkSubmitCmd.concat(" %s %s".format(PIPELINE, SubmitJobJson.format(sparkDeployMode)))
+      } else {
+        sparkSubmitCmd
+      }
+
+      cmds.writeScript(sparkCommand, scriptPath, executionDir)
+      scriptPath.addPermission(PosixFilePermission.OWNER_EXECUTE)
+
+    } catch {
+      case ex: Exception =>
+        log.error(ex, "Failed to prepare task: " + ex.getMessage)
+        executionResponse success FailedNonRetryableResponse(jobDescriptor.key, ex, None)
+    }
+  }
+
+  private def checkForSparkClusterMode(deployMode: String, master: String): Boolean = {
```

Comment:
can't this be calculated once and cached?  it's currently called in three different spots